File: prototypes/research-assistant/python/app/api/research.py
"""Research route handlers."""

import uuid

from fastapi import APIRouter

from app.models.schemas import (
    ResearchRequest,
    ResearchResult,
    ResearchStatus,
    ResearchStep,
    Source,
)

router = APIRouter()

_results: dict[str, dict] = {}


@router.post("/research", response_model=ResearchResult)
async def start_research(request: ResearchRequest):
    """Start a research session."""
    research_id = str(uuid.uuid4())
    trace_id = str(uuid.uuid4())

    steps = [
        ResearchStep(step=1, action="search", content=f"Searching for: {request.question}"),
        ResearchStep(step=2, action="analyze", content="Analyzing search results"),
        ResearchStep(step=3, action="synthesize", content="Synthesizing findings"),
    ]

    sources = [
        Source(
            title="Example Source",
            url="https://example.com",
            excerpt="Relevant information found here.",
        ),
    ]

    result = ResearchResult(
        id=research_id,
        question=request.question,
        steps=steps,
        answer=f"Based on research, here is the answer to: {request.question}",
        sources=sources,
        trace_id=trace_id,
    )

    _results[research_id] = {"status": "completed", "steps": len(steps)}
    return result


@router.get("/research/{research_id}/status", response_model=ResearchStatus)
async def get_research_status(research_id: str):
    """Get status of a research session."""
    info = _results.get(research_id, {"status": "not_found", "steps": 0})
    return ResearchStatus(
        id=research_id,
        status=info["status"],
        steps_completed=info["steps"],
    )

File: prototypes/research-assistant/typescript/src/api/research.ts
import { Hono } from "hono";
import { ResearchRequest } from "../schemas/index.js";

export const researchRouter = new Hono();

const results: Map<string, { status: string; steps: number }> = new Map();

researchRouter.post("/research", async (c) => {
  const body = await c.req.json();
  const parsed = ResearchRequest.safeParse(body);

  if (!parsed.success) {
    return c.json(
      { error: "Invalid request", details: parsed.error.issues },
      400,
    );
  }

  const researchId = crypto.randomUUID();
  const traceId = crypto.randomUUID();

  const steps = [
    {
      step: 1,
      action: "search",
      content: `Searching for: ${parsed.data.question}`,
    },
    { step: 2, action: "analyze", content: "Analyzing search results" },
    { step: 3, action: "synthesize", content: "Synthesizing findings" },
  ];

  const sources = [
    {
      title: "Example Source",
      url: "https://example.com",
      excerpt: "Relevant information found here.",
    },
  ];

  results.set(researchId, { status: "completed", steps: steps.length });

  return c.json({
    id: researchId,
    question: parsed.data.question,
    steps,
    answer: `Based on research, here is the answer to: ${parsed.data.question}`,
    sources,
    trace_id: traceId,
  });
});

researchRouter.get("/research/:researchId/status", (c) => {
  const researchId = c.req.param("researchId");
  const info = results.get(researchId) ?? { status: "not_found", steps: 0 };

  return c.json({
    id: researchId,
    status: info.status,
    steps_completed: info.steps,
  });
});
